eXist-db/public-repo
# version 2.0

Closes #46 and #49.

BREAKING CHANGES

- Packages are now stored outside the `public-repo` collection, in a sibling `public-repo-data` collection. Packages are no longer deleted when upgrading the public repo app.
- To prevent permissions issues, uploading and publishing apps is limited to members of the `repo` group.
- The `public-repo-data` collection is split into 3 sub-collections: `packages`, `icons`, and `metadata`. 
- Despite the changes in storage, no public API changes have been introduced: the HTML listings, download URLs, API endpoints, etc. are unchanged.
- The metadata files have been renamed from `packages.xml` (introduced in #52) to `raw-packages.xml` (with the structure `/raw-packages/package`) and from `apps.xml` to `package-groups.xml` (with the structure `/package-groups/package-group`). 
- The app now requires eXist 5. 
- The app's dependency on the crypto library has been removed.
- Due to the breaking changes, I've gone ahead and incremented the version number in this PR to 2.0.0. 

To aid in upgrading existing installations to this location and format, the PR includes a script that performs the upgrade from the v1 format to the v2 format—which must be done BEFORE upgrading the app itself. I'll send out instructions on how to use this script to exist-open and Slack when the PR is merged.

I've tested the PR extensively with eXist 5.2.0, but DO NOT MERGE BEFORE SOMEONE WHO HAS TESTED IT APPROVES. (~~The app is broken in 5.3.0-SNAPSHOT due to its dependency on the EXPath Crypto library, which is broken.~~ The app is compatible with 5.3.0-SNAPSHOT since it no longer depends on the Crypto library.)

To test:

1. install public-repo v1.0.3 or v1.1.0 on an eXist 5.1.1 system (the version running on the exist-db.org server now) or newer.
2. upload a representative sample of xars and publish them (from admin.html, which triggers an AJAX request served by `upload.xq`); filenames of uploaded xars should appear in the "Uploaded Files" section 
3. run the upgrade script in `modules/upgrade-to-v2-storage.xq`
4. install this PR
5. the xar listing on the landing page, package-specific listings on interior pages, and all endpoints should respond like normal
    - landing page xar listing: http://localhost:8080/exist/apps/public-repo
    - package-specific listings, with version parameters: http://localhost:8080/exist/apps/public-repo/packages/semver-xq.html?eXist-db-min-version=4.7.0
    - xar download service, served by `controller.xql` which forwards requests to the data collection: http://localhost:8080/exist/apps/public-repo/public/eXide-2.4.10.xar
    - icon download service, served by `controller.xql`, which forwards requests to the data collection: http://localhost:8080/exist/apps/public-repo/public/eXide-2.4.10.png
    - atom feed, served by `modules/feed.xq`: http://localhost:8080/exist/apps/public-repo/feed.xml
    - version-specific listing needed for packageservice, served by list.xq: http://localhost:8080/exist/apps/public-repo/public/apps.xml?version=5.2.0 (and lower versions to confirm correct filtering)
    - package retrieval service needed by eXist build tools and the `repo:install-and-deploy` function, served by `modules/find.xq`: http://localhost:8080/exist/apps/public-repo/find?abbrev=eXide&processor=5.2.0 (append "&info=yes" for a debug version)
    - .zip file extension download service, served by `modules/download-xar-zip.xq`: http://localhost:8080/exist/apps/public-repo/public/eXide-2.4.10.xar.zip
    - rebuild package metadata service, served by update.xq
6. upload new xars to confirm the can be published.
7. check the various pages and endpoints again
6. for good measure, reinstall the PR and confirm that xars are not deleted.

Thanks to @line-o for making public-repo scalable again with #52. This was the breakthrough that inspired me to dig deeper into this problem, at long last.
